mitosheet/mitosheet/code_chunks/code_chunk_utils.py

```diff
@@ -56,13 +51,63 @@ def optimize_code_chunks(all_code_chunks: List[CodeChunk]) -> List[CodeChunk]:
     if len(all_code_chunks_reversed) == 1:
         code_chunks_list.append(all_code_chunks_reversed[0])
 
-    # TODO: we could combine_left here? This would allow us to 
-    # express overwriting conditions very naturally for deleting
-    # dataframes...
+    return optimized, code_chunks_list
+
+def optimize_code_chunks_combine_left(code_chunks_to_optimize: List[CodeChunk]) -> Tuple[bool, List[CodeChunk]]:
+    """
+    Given a list of code chunks, willattempt to optimize them left
+    with a single pass.
+    """
+
+    code_chunks_to_optimize = copy(code_chunks_to_optimize)
+    code_chunks_list: List[CodeChunk] = []
+
+    optimized = False
+    while len(code_chunks_to_optimize) >= 2:
+        later_code_chunk = code_chunks_to_optimize.pop()
+        earlier_code_chunk = code_chunks_to_optimize.pop()
+
+        combined_chunk = later_code_chunk.combine_left(earlier_code_chunk)
+
+        if combined_chunk is not None:
+            # If we can combine the two chunks into one, take that
+            optimized = True
+            code_chunks_to_optimize.append(combined_chunk)
+        else:
+            # If we cannot combine the two chunks, take the first chunk, 
+            # and reset the second one to the new first chunk for the next loop
+            code_chunks_list.append(later_code_chunk)
+            code_chunks_to_optimize.append(earlier_code_chunk)            
```

Comment:
This comment is confusing because the `first chunk` is referring to the `later_code_chunk`.